modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/AddCatalogWorkflowOperationHandler.java

```diff
@@ -0,0 +1,227 @@
+
+  /** The mimetype of the added catalogs */
+  private static final MimeType CATALOG_MIME_TYPE = MimeType.mimeType("text", "xml");
+
+  /** The workspace, where the catalog files are put. */
+  private Workspace workspace;
+
+  /**
+   * Sets the workspace to use.
+   *
+   * @param workspace
+   *          the workspace
+   */
+  public void setWorkspace(Workspace workspace) {
+    this.workspace = workspace;
+  }
+
+  @Override
+  public WorkflowOperationResult start(WorkflowInstance wInst, JobContext context)
+      throws WorkflowOperationException {
+    // get Workflow configuration
+    String catalogName = getCfgValueFromOp(wInst, CFG_KEY_CATALOG_NAME, false);
+    String catalogPath = getCfgValueFromOp(wInst, CFG_KEY_CATALOG_PATH, false);
+    String catalogTags = getCfgValueFromOp(wInst, CFG_KEY_CATALOG_TAGS, true);
+    MediaPackageElementFlavor    catalogFlavor = parseFlavor(
+                           getCfgValueFromOp(wInst, CFG_KEY_CATALOG_FLAVOR, false));
+    CatalogTypeCollisionBehavior collBehavior  = parseCollisionBehavior(
+                           getCfgValueFromOp(wInst, CFG_KEY_CATALOG_TYPE_COLLISION_BEHAVIOR, false));
+
+    MediaPackage mp = wInst.getMediaPackage();
+
+    // if CatalogType is already part of the MediaPackage handle special cases
+    if (doesCatalogFlavorExist(catalogFlavor, mp.getCatalogs())) {
+      if (collBehavior == CatalogTypeCollisionBehavior.FAIL) {
+        throw new WorkflowOperationException("Catalog Type already exists and 'fail' was specified");
+      }
+      else if (collBehavior == CatalogTypeCollisionBehavior.SKIP) {
+        // don't add the Catalog
+        return createResult(mp, Action.CONTINUE);
+      }
+    }
+
+    // 'upload Catalog' to workspace
+    File   catalogFile = new File(catalogPath);
+    String catalogId   = UUID.randomUUID().toString();
+    URI    catalogURI  = null;
+    try (InputStream catalogInputStream = FileUtils.openInputStream(catalogFile)) {
+      catalogURI = workspace.put(mp.getIdentifier().toString(), catalogId,
+              catalogName, catalogInputStream);
+    }
+    catch (IOException e) {
+      throw new WorkflowOperationException(e);
+    }
+
+    // add Catalog to MediaPackage (and set Properties)
+    MediaPackageElement mpe = mp.add(catalogURI, MediaPackageElement.Type.Catalog, catalogFlavor);
+    mpe.setIdentifier(catalogId);
+    mpe.setMimeType(CATALOG_MIME_TYPE);
+    if (catalogTags != null) {
+      for (String tag : catalogTags.split(",")) {
+        mpe.addTag(tag);
+      }
+    }
+
+    return createResult(mp, Action.CONTINUE);
+  }
+
+  /**
+   * Checks whether the catalogFlavor exists in the array of catalogs
+   *
+   * @param catalogFlavor
+   * @param catalogs
+   * @return true, if the catalogFlavor exists in the array of catalogs, else false
+   */
+  private boolean doesCatalogFlavorExist(MediaPackageElementFlavor catalogFlavor, Catalog[] catalogs) {
+    List<Catalog> ctls = Arrays.asList(catalogs);
```

Comment:
In theory, you could write just...

`
return Arrays.asList(catalogs).stream().anyMatch(catalog -> catalogFlavor.matches(catalog.getFlavor()));
`

here.